### ISSUE TITLE
State of the Project

### ISSUE THREAD

**Issue by mattbdean** [mattbdean/JRAW#187]
# Current State

I'd like to start out by thanking everyone who's contributed to this project. Your help has made JRAW into a success!

As many of you might have probably noticed, I have been slacking hard on addressing issues and looking at pull requests. I recently got a new job in the webdev/JavaScript world and most of my open source work has been there as well. Taking a look at the library, its source code could use an overhaul.

 - **Second-class OAuth2 support**: JRAW was created before OAuth2 was the standard for authenticating reddit apps and its design still reflects that. A user new to this library might create a RedditClient and immediately try to send a request and wonder why it fails before either diving into the Javadoc, source code, or wiki.
 - **Testing**: JRAW's tests are mostly unintuitive and prone to failure. Take a look at [HttpAdapterTest](https://github.com/thatJavaNerd/JRAW/blob/4f2346aad873983f6f84e9858ae3936ead7c2275/src/test/java/net/dean/jraw/test/HttpAdapterTest.java) for example. Anybody unfamiliar with TestNG is going to confused when trying to add a new test or understand existing tests.
 - **Public API**: JRAW's public API is verbose and relies a lot on [managers](https://github.com/thatJavaNerd/JRAW/tree/master/src/main/java/net/dean/jraw/managers) which, while still works, could be designed a lot better.
 - **Internals**: I created JRAW before I fully understood what Jackson Databind could be used for. Every JsonModel has a dozen or so basic getters which simply look up the value of a JsonNode. Jackson is insanely good at binding JSON data to a POJO, and not utilizing that functionality drastically increases the amount of boilerplate code.
 - **`:endpoints`**: The endpoints subproject creates ENDPOINTS.md, Endpoints.java, and updates the "api-coverage" badge in the README. It works, but my code is unsavory.

# v1.0.0

**JRAW v1.0.0 will be written in [Kotlin](https://kotlinlang.org/)**.

Kotlin is JetBrain's succinct, statically-typed JVM-based language which had its first public release in 2012. Null-safety built in to the language and it supports first class functions (lambdas).

JRAW v1.0.0 will attempt to address the issues I mentioned above. You can check out the current state of the rewrite on the [kotlin](https://github.com/thatJavaNerd/JRAW/tree/kotlin) branch.

 - **OAuth2 now has first class support.** RedditClients cannot be created without authentication.
 - **Testing will be done with [Spek](http://spekframework.org/) and [Expekt](https://github.com/winterbe/expekt)**. If you've ever used [mocha](https://mochajs.org/) and [chai](http://chaijs.com/) you should feel right at home.
 - **The public API will be rewritten to be clear and concise.** This means no more semi-[god objects](https://en.wikipedia.org/wiki/God_object).
 - **Models will be created with ~~Jackson Databind~~ Moshi.**

## "But what does this mean for me?"

If you are using JRAW in your projects, don't worry. Kotlin compiles to JVM bytecode and [works well with Java](https://kotlinlang.org/docs/reference/java-to-kotlin-interop.html). **This release (like the ones before it) will simply mean API changes**.

(and yes, JRAW will keep it's name)

**Comment by Saketme** [comment]
Good to hear that this project will still be actively maintained. 

May I suggest not using Jackson directly as the underlying data-set for POJOs and only using it for serializing/deserializing them? Because in its current form, JRAW enforces client apps to also use Jackson for serializing/deserializing the POJOs instead of their choice of libraries. Passing them around as parcelables is also not possible.

**Comment by thatJavaNerd** [comment]
@Saketme 

>JRAW enforces client apps to also use Jackson for serializing/deserializing the POJOs instead of their choice of libraries

What do you mean? Are you talking about how [JsonModel](https://github.com/thatJavaNerd/JRAW/blob/master/src/main/java/net/dean/jraw/models/JsonModel.java) requires a Jackson JsonNode in its constructor?

>Passing them around as parcelables is also not possible.

What would have to be done to make this possible?

**Comment by Saketme** [comment]
> What do you mean? Are you talking about how JsonModel requires a Jackson JsonNode in its constructor?

Yes, as a result the only way to create their new objects is by using Jackson. I think this is slightly bad because you're tightly coupling Jackson with JRAW. 

Additionally, since methods like `AccountManager#save(PublicContribution)`, `InboxManager#setRead(Message)` require the entire object instead of just their full-names, it's difficult to pass them around. For example, I was saving their data-node's JSON and bundling them together with `PendingIntents` for message notifications. This way the objects can be supplied to the `InboxManager` when they're marked as read. I think this would have been easier if they implemented `Parcelable`.

At the same time, I'm not sure if this should be done because JRAW is not an Android library. It's a plain Java library, right?

One hackery that I'm doing in my app to workaround this restriction is wrapping the full-name of `Things` in another object that extends `Thing` so that they can be supplied to the managers. This is working for me so far.

**Comment by thatJavaNerd** [comment]
> At the same time, I'm not sure if this should be done because JRAW is not an Android library. It's a plain Java library, right?

You are correct, JRAW is a plain Java library. There is a very small and old [extension library](https://github.com/thatJavaNerd/JRAW-Android) that I wrote a while back that basically generates a UserAgent for you and stores your tokens in SharedPreferences.

I realize that this library is going to be used mostly in Android apps since most scripts are written in python. My goal for v1.0.0 for it to be viable in any JVM environment.

>> What do you mean? Are you talking about how JsonModel requires a Jackson JsonNode in its constructor?

> Yes, as a result the only way to create their new objects is by using Jackson. I think this is slightly bad because you're tightly coupling Jackson with JRAW.

I completely agree with you. My rationale for doing so was that reddit is constantly adding new properties to its models, and if developers wanted to get a property that a JsonModel didn't have a getter for, they wouldn't be able to without access to its backing data node. Ultimately, I think decoupling Jackson from the models would be the better design choice. Worst case scenario developers send the request themselves instead of using JRAW's API.

```kotlin
val foo: Foo = redditClient.sendRequestThatReturnsFoo()
val barProperty: String = foo.bar
// versus
val foo: JsonNode = redditClient.request { it.path("/api/v1/foo") }.json
val barProperty: String = fooData.get("bar").asText()
```

> Additionally, since methods like AccountManager#save(PublicContribution), InboxManager#setRead(Message) require the entire object instead of just their full-names, it's difficult to pass them around.

I did this to make sure developers wouldn't go around passing an ID instead of a fullname. If you take a look at [`AccountManager.setSaved()`](https://github.com/thatJavaNerd/JRAW/blob/36aaf752476b5989383d8a96c6761c3635114828/src/main/java/net/dean/jraw/managers/AccountManager.java#L156), it only uses the `getFullName()` method from the PublicContribution. The rewrite will definitely require just the fullname String instead of the entire object for methods like this.

See #188 for more

**Comment by Saketme** [comment]
> My rationale for doing so was that reddit is constantly adding new properties to its models, and if developers wanted to get a property that a JsonModel didn't have a getter for, they wouldn't be able to without access to its backing data node.

Ah, that is also a valid reason. I must say that access to the underlying data-node is indeed helpful, because I've manually extracted information from data models this way in some places. I'm now not sure what's the best way to achieve both.

**Comment by ccrama** [comment]
Have you considered making JRAW (or a version of JRAW) [Realm ](https://realm.io/)friendly? @fbis251  and I were planning on re-writing JRAW using Realm objects, but we didn't get very far with actually doing this as Realm doesn't support polymorphism, which makes a direct port pretty difficult (the 'Thing' class, for example, cannot have child classes that inherit it if it is a Realm object). The whole Paginator base of JRAW would have to be generalized to use Realm objects as well, but if v1 was not totally polymorphic, it would make a port much easier. 

IMO this would make Android integration and data persistence much easier to do! Just an idea, might be something to consider. If not, I am currently planning a rewrite of Slide's backend, and am going to make some sort of interface that bridges Realm with JRAW

**Comment by thatJavaNerd** [comment]
@Saketme Apparently we can use `@JsonAnyGetter` for any unknown properties, so the end result would look something like this:

```kotlin
foo: Foo = ...
unknownProperty: Any = foo.get("bar");
```

**Comment by fbis251** [comment]
Would you consider using Retrofit2 for the API calls? This will help with several things such as providing a JSON converter independent way of providing access to the reddit API and would allow for simpler OAuth2 token requests

Firstly, the API is defined via an interface. Any user of your library will be able to open this interface and immediately see all the API calls available to them, with their return types and expected parameters. This will allow the user to be able to make the API calls directly themselves if needed, but will still allow JRAW to provide Paginators and other classes that wrap these calls in a more consistent manner.

Currently these JSON libraries are supported by Retrofit2:
https://github.com/square/retrofit/wiki/Converters

There is also call adapter support that will allow adding RxJava1/2 support to your library without requiring it as a compile-time dependency for the library user:
https://github.com/square/retrofit/wiki/Call-Adapters

Although only OkHttp3 is supported as a client, Retrofit + OkHttp3 provides an easy way to handle authentication via OAuth2. I don't see that only OkHttp is supported as a downside as most recent Android versions are using it as the default HTTP client
https://futurestud.io/tutorials/oauth-2-on-android-with-retrofit

Basically for authentication using OkHttp3 you can use an Authenticator and an Interceptor, you can make the client request new tokens as needed without the library user needing to handle this themselves. This all happens in whatever thread the original reddit API request was made which means that threading also will not be an issue for OAuth refresh requests.

I have been prototyping a stripped-down reddit client using Retrofit2 for one of my projects and believe that JRAW can benefit greatly from it. The only issue I see is accessing fields not explicitly defined in your POJOs. I'm not sure if GSON can do it, I believe Moshi can, and I know Jackson can do it.

**Comment by thatJavaNerd** [comment]
See my comment [here](https://github.com/thatJavaNerd/JRAW/issues/188#issuecomment-308261514) on Retrofit

**Comment by mattbdean** [comment]
Just about ready for v1.0.0, so I'm closing this issue
